### PR TITLE
added utilities to Cigar*

### PR DESCRIPTION
**Pull request by lindenb**
I've added a few functions to Cigar* . I'm working with split reads and I often need to know if the cigar is clipped, what is the nature if the 5' or 3' operator ,etc ...

in **CigarOperator**:
* isClipping
* isPadding
* isIndel
* isIndelOrSkippedRegion
* isAlign

in **Cigar**
* cigarElements is final and so it is never nulll
* now implements Iterable<CIgarElement>
* added static builder from a List of CigarOperator (it cannot be a constructor because a similar constructor exists)
* added 'containsOperator', isLeftClipped, isRightClipped, isClipped
* prevent IDE warnings in 'switch/case' by adding 'default'
* get first/last cigarElement

in **CigarElement**:
* added toString

in **CigarTest**:

added one test for the above functions.

### CODE REVIEW

File: src/java/htsjdk/samtools/Cigar.java [hunk 1]
```diff
@@ -254,15 +259,69 @@ public boolean equals(final Object o) {
 
         final Cigar cigar = (Cigar) o;
 
-        if (cigarElements != null ? !cigarElements.equals(cigar.cigarElements) : cigar.cigarElements != null)
-            return false;
+        return cigarElements.equals(cigar.cigarElements);
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/Cigar.java]: why did you remove the null protections?

Comment (lindenb) [inline src/java/htsjdk/samtools/Cigar.java]: cigarElements is never null:

```java
private final List<CigarElement> cigarElements = new ArrayList<CigarElement>();
```

File: src/java/htsjdk/samtools/Cigar.java [hunk 2]
```diff
@@ -254,15 +259,69 @@ public boolean equals(final Object o) {
...
+    }
+    
+    /** build a new cigar from a list of cigar operators */
+    public static Cigar fromCigarOperators(final List<CigarOperator> cigarOperators) {
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/Cigar.java]: what is the use-case here?, could you explain in the javadoc?

File: src/java/htsjdk/samtools/Cigar.java [hunk 3]
```diff
@@ -254,15 +259,69 @@ public boolean equals(final Object o) {
...
+    
+    /** shortcut to <code>getCigarElements().iterator()</code> */
+    @Override
+    public Iterator<CigarElement> iterator() {
+        return this.getCigarElements().iterator();
+    }
+    
+    /** returns true if the cigar string contains the given operator */
+    public boolean containsOperator(final CigarOperator operator) {
+        for(final CigarElement cigarElement : this.cigarElements) {
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/Cigar.java]: care to java8-ify this function?

cigarElements.stream().anyMatch(el->el.getOperator()==op);

File: src/java/htsjdk/samtools/Cigar.java (hunk 1 above)
Comment (yfarjoun) [inline src/java/htsjdk/samtools/Cigar.java]: I see. cigarElements should never be null. OK.

File: src/java/htsjdk/samtools/CigarOperator.java [hunk 4]
```diff
@@ -179,6 +179,31 @@ public static byte enumToCharacter(final CigarOperator e) {
...
+        return this == I || this == D;
+    }
+
+    /** Returns true if the operator is a Skipped Region Insertion or Deletion operator */
+    public boolean isIndelOrSkippedRegion() {
+        return this == N || isIndel();
+    }
+
+    /** Returns true if the operator is a M, a X or a EQ */
+    public boolean isAlign() {
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/CigarOperator.java]: should be a noun, like the others, not a verb. isAlignment ?

File: src/java/htsjdk/samtools/Cigar.java [hunk 5]
```diff
@@ -254,15 +259,69 @@ public boolean equals(final Object o) {
...
-        if (cigarElements != null ? !cigarElements.equals(cigar.cigarElements) : cigar.cigarElements != null)
-            return false;
+        return cigarElements.equals(cigar.cigarElements);
+    }
+    
+    /** build a new cigar from a list of cigar operators */
+    public static Cigar fromCigarOperators(final List<CigarOperator> cigarOperators) {
+        if( cigarOperators == null ) throw new IllegalArgumentException("cigarOperators is null");
+        final List<CigarElement> cigarElementList = new ArrayList<>();
+        int i=0;
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/Cigar.java]: spaces in this function

File: src/java/htsjdk/samtools/Cigar.java [hunk 6]
```diff
@@ -254,15 +259,69 @@ public boolean equals(final Object o) {
...
+    }
+    
+    /** returns the first cigar element */
+    public CigarElement getFirstCigarElement() {
+        return isEmpty() ? null : this.cigarElements.get(0); 
+    }
+    
+    /** returns the last cigar element */
+    public CigarElement getLastCigarElement() {
+        return isEmpty() ? null : this.cigarElements.get(this.numCigarElements() -1);
```
Comment (yfarjoun) [inline src/java/htsjdk/samtools/Cigar.java]: -<space>1

**Comment by yfarjoun** [comment]
minor comments. looks good!

back to you @lindenb

**Comment by lindenb** [comment]
@yfarjoun  I've updated the code + squashed.

by the way : in https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/Cigar.java#L192 the old code only looks at 'I' and 'D' but not 'N' as it is declared.

```java
                // There should be an M or P operator between any pair of IDN operators
                if (isInDelOperator(op)) {
(...)

  private static boolean isInDelOperator(final CigarOperator op) {
        return op == CigarOperator.I || op == CigarOperator.D;
    }
```

I kept the old code but I wonder if you shouldn't consider 'op == CigarOperator.N' too ?

**Comment by yfarjoun** [comment]
If the current code doesn't do the right thing (please add a test!) then,
by all means, I'd love if you could add a fix!

can go in a separate PR though.

this one is fine! thanks! :+1:

**Comment by lindenb** [comment]
@yfarjoun ok, I'll do this in another PR.
